VMODL API for multi-tenancy support

shaominchen opened the pull request:
The implementation (to integrate with auth_api) has not been implemented yet. Please review the API signatures first.

msterin commented on esx_service/vmodl/DatastoreAccessPrivilege.py:
```diff
@@ -0,0 +1,93 @@
+
+   @JavaDocs(parent=_name, docs =
+   """
+   Max volume size allowed on this datastore
```
need units. I suggest the attribute name to have "in_bytes" or whatever at the end. Actually, in_Mib should be fine

shaominchen replied:
This PR will include VMODL definitions only. In the next PR I will provide the implementation along with the unit tests.

Will add _in_MiB suffix.

msterin commented on esx_service/vmodl/DatastoreAccessPrivilege.py:
```diff
@@ -0,0 +1,93 @@
+   """
+   )
+   @Attribute(parent=_name, typ="string")
+   def usage_quota(self):
```
_in_MiB ?

shaominchen replied:
Sure, will add _in_MiB suffix.

msterin commented on esx_service/vmodl/DatastoreAccessPrivilege.py:
```diff
@@ -0,0 +1,93 @@
+
+   @JavaDocs(parent=_name, docs =
+   """
+   Datastore name
```
what if the datastore name changes (and url stays the same) ? Should we keep url here  ?

shaominchen replied:
Good catch! I will use datastore URL instead. BTW, this issue already exists in the existing logic. I have filed a tracking issue: #782.

msterin commented on esx_service/vmodl/Tenant.py:
```diff
@@ -0,0 +1,210 @@
+
+   @JavaDocs(parent=_name, docs =
+   """
+   Tenant name
```
Any limitations ? Length ? Char set ? Can it be unicode ?

shaominchen replied:
I will specify the length of the string. AFAIK, VMODL string supports unicode on all different language bindings: C++, Java, Python, etc. But this is implicit and we don't usually document this in the JavaDoc.

msterin commented on esx_service/vmodl/Tenant.py:
```diff
@@ -0,0 +1,210 @@
+   """
+   )
+   @Method(parent=_name, wsdlName="AddVMs")
+   @Param(name="vms", typ="string[]")
```
a question abut all *VMs() operations: should we use vim.VirtualMachine MoRef instead of strings here ?

shaominchen replied:
Will use MoRef consistently for all references of VMs and Datastores.

msterin commented on esx_service/vmodl/TenantManager.py:
```diff
@@ -0,0 +1,103 @@
+   """
+   )
+   @Method(parent=_name, wsdlName="RemoveTenant")
+   @Param(name="name", typ="string")
```
vim.vcs.Tenant instead of string here ?

shaominchen replied:
Since all Tenant related APIs use tenant name as the unique key, I think it makes sense to also use name as the key for this API. What do you think?

msterin commented on esx_service/vmodl/TenantManager.py:
```diff
@@ -0,0 +1,103 @@
+
+   @JavaDocs(parent=_name, docs=
+   """
+   Query Tenants for the given name. Return all Tenants if name is null.
```
will we support globbing (e.g. ? *) If yes needs to be documented

shaominchen replied:
I've thought about this. Globbing is typically used by shells for matching file and directory names using wildcards. Even though our Tenant name does map to a directory in the backend, this implementation detail should not be exposed to customers. Regular expressions are actually more commonly used in commands / functions for pattern matching in string text. However, supporting regex seems an overkill for our current requirements.

So, for now I think we can keep it simple, i.e. not supporting globbing or regex. In the future if there are real requirements for this, we can add this feature without breaking backward compatibility.

msterin commented on esx_service/vmodl/TenantManager.py:
```diff
@@ -0,0 +1,103 @@
+   @Param(name="new_name", typ="string", flags=F_OPTIONAL)
+   @Param(name="new_description", typ="string", flags=F_OPTIONAL)
+   @Return(typ="void")
+   def RemoveTenant(self, name):
```
needs to be ModifyTenant

shaominchen replied:
Copy-paste error. Will fix.

msterin commented on esx_service/vmodl/VsphereContainerService.py:
```diff
@@ -0,0 +1,65 @@
+
+
+# _VERSION = newestVersions.Get("vim")
+_VERSION = 'vim.version.version10'
```
I think common modules import and version should be imported from a single file instead of copy-pasted to each module

shaominchen replied:
Make sense, will do.

shaominchen replied:
I've decided to merge all classes into one single file/module as this seems more consistent with Python style. So no need to introduce a common module anymore.

shaominchen commented on esx_service/vmodl/Tenant.py:
```diff
@@ -0,0 +1,210 @@
+   This class provides operations to manage VMs and privileges associated with a tenant.
+   '''
+
+   _name = "vim.vcs.TenantManager"
```
Thanks for catching this. Will fix.

shaominchen commented on esx_service/vmodl/Tenant.py:
```diff
@@ -0,0 +1,210 @@
+   @Method(parent=_name, wsdlName="RemoveVMs")
+   @Param(name="vms", typ="string[]")
+   @Return(typ='void')
+   def RemoveVMsFromTenant(self, vms):
```
Will fix.

shaominchen commented:
@msterin @lipingxue 

I've updated the PR based on your comments. Also redesigned some APIs. Major changes since last update:
1. Add detailed comments for all classes, methods, fields, parameters, return values and exceptions.
2. Add exception handling for all VMODL APIs
3. Use MoRef for all Datastore and VM references
4. Merge all classes into one single file/module

Exception handling is a major missing part that we haven't discussed before. I plan to introduce one or more exceptions to handle all different error conditions. Maybe we should have a discussion about this after everyone is back.

shaominchen commented on esx_service/vmodl/VsphereContainerService.py:
```diff
@@ -0,0 +1,377 @@
+   )
+   @Method(parent=_name, wsdlName="RemoveTenant")
+   @Param(name="name", typ="string")
+   @Param(name="remove_volumes", typ="boolean")
```
Will address this.

shaominchen commented on esx_service/vmodl/VsphereContainerService.py:
```diff
@@ -0,0 +1,377 @@
+
+   @JavaDocs(parent=_name, docs =
+   """
+   Tenant is an abstraction of a group of VMs and the associated access
```
Will clarify a little bit.

shaominchen commented:
@lipingxue Addressed your comments.

msterin commented on esx_service/vmodl/VsphereContainerService.py:
```diff
@@ -0,0 +1,400 @@
+        faults=["vim.fault.NotFound"])
+   @Param(name="datastore", typ="vim.Datastore", flags=F_OPTIONAL)
+   @Return(typ='vim.vcs.storage.DatastoreAccessPrivilege[]')
+   def GetPrivileges(self, datastore):
```
nit: it feels strange that AddPrivilige is singular but GetPrivileges is plural. I suggest to stick with singular everywhere.

shaominchen replied:
I'm actually following existing VSAN VMODL style. Basically, if the return type is an array, the name will be plural, for example: GetTenants, GetVMs, GetPrivileges. Or if the input parameter is an array, the name is also plural: AddVMs, RemoveVMs, ReplaceVMs.

Please let me know if you still have concern about this. I can change it in the next PR if needed.

lipingxue commented:
LGTM to me.
